src/host/multiple_ledger.h

```diff
@@ -0,0 +1,815 @@
+      max_read_cache_size(max_read_cache_size)
+    {
+      if (chunk_threshold == 0 || chunk_threshold > max_chunk_threshold_size)
+      {
+        throw std::logic_error(fmt::format(
+          "Error: Ledger chunk threshold should be between 1-{}",
+          max_chunk_threshold_size));
+      }
+
+      if (fs::is_directory(ledger_dir))
+      {
+        // If the ledger directory exists, recover ledger files from it
+        for (auto const& f : fs::directory_iterator(ledger_dir))
+        {
+          files.push_back(
+            std::make_shared<LedgerFile>(ledger_dir, f.path().filename()));
+        }
+
+        files.sort([](
+                     const std::shared_ptr<LedgerFile>& a,
+                     const std::shared_ptr<LedgerFile>& b) {
+          return a->get_last_idx() < b->get_last_idx();
+        });
+
+        last_idx = get_latest_file()->get_last_idx();
+
+        for (auto f = files.begin(); f != files.end();)
+        {
+          if ((*f)->is_compacted())
+          {
+            compacted_idx = (*f)->get_last_idx();
+            auto f_ = f;
+            f++;
+            files.erase(f_);
+          }
+          else
+          {
+            f++;
+          }
+        }
+
+        // Continue writing at the end of last file only if that file is not
+        // complete
+        if (files.size() > 0 && !files.back()->is_complete())
+        {
+          require_new_file = false;
+        }
+        else
+        {
+          require_new_file = true;
+        }
+      }
+      else
+      {
+        if (!fs::create_directory(ledger_dir))
+        {
+          throw std::logic_error(fmt::format(
+            "Error: Could not create ledger directory: {}", ledger_dir));
+        }
+        require_new_file = true;
+      }
+    }
+
+    MultipleLedger(const MultipleLedger& that) = delete;
+
+    const std::vector<uint8_t> read_entry(size_t idx)
+    {
+      auto f = get_file_from_idx(idx);
+      if (f == nullptr)
+      {
+        return {};
+      }
+      return f->read_entry(idx);
+    }
+
+    const std::vector<uint8_t> read_framed_entries(size_t from, size_t to)
+    {
+      if ((from <= 0) || (to > last_idx) || (to < from))
+      {
+        return {};
+      }
+
+      std::vector<uint8_t> entries;
+      size_t idx = from;
+      while (idx <= to)
+      {
+        auto f_from = get_file_from_idx(idx);
+        if (f_from == nullptr)
+        {
+          return {};
+        }
+        auto to_ = std::min(f_from->get_last_idx(), to);
+        auto v = f_from->read_framed_entries(idx, to_);
+        if (v.size() == 0)
+        {
+          return entries;
+        }
+        entries.insert(
+          entries.end(),
+          std::make_move_iterator(v.begin()),
+          std::make_move_iterator(v.end()));
+        idx = to_ + 1;
+      }
+
+      return entries;
+    }
+
+    size_t write_entry(const uint8_t* data, size_t size, bool committable)
+    {
+      if (require_new_file)
```

Comment:
Haven't looked at this in detail, but it seems risky to be juggling this bool between functions. Can we instead create a new file at any point where we'd be setting this? So we create new files eagerly, rather than lazily.

Reply:
As discussed, creating a new file whenever the threshold is passed would complicate the truncation logic which I think we should avoid. Leaving the bool in for now but happy to revisit in the future (e.g., when implementing #1276).